---
title: "CASCADE HER2 Gene Expression Analysis"
author: "Thomas Conway"
date: "2020-05-11"
output:
  html_document:
    toc: true
  pdf_document:
    keep_tex: true
---

# CASCADE HER2+ Gene Expression Analysis

The CASCADE project is a study across several cancer types, in which
in addition to any samples taken at diagnosis and during treatment,
at death, a rapid autopsy is performed to sample metastatic leisions
that may not be accessable otherwise. Among the various assays used, are
whole genome sequencing (WGS), and whole transcriptome sequencing (WTS).

Here we present the differential gene expression (DGE) analysis of the
4 metastatic sites sampled from patient CAS0057, who was diagnosed with
HER2+ breast cancer.

## Preamble

```{r, echo=FALSE}
library(data.table)
library(ggplot2)
library(scales)

library(EnsDb.Hsapiens.v75)
library(edgeR)

gene.names <- select(EnsDb.Hsapiens.v75, keys=keys(EnsDb.Hsapiens.v75, keytype="GENEID"), keytype="GENEID", columns=c("GENEID", "GENENAME"))
gene.names <- data.table(ensembl.id = gene.names$GENEID, name = gene.names$GENENAME)
setkey(gene.names, 'ensembl.id')

commaize <- label_comma()
percentize <- label_percent()

pam50.genes <- c("UBE2T", "BIRC5", "NUF2", "CDC6", "CCNB1",
                 "TYMS", "MYBL2", "CEP55", "MELK", "NDC80",
                 "RRM2", "UBE2C", "CENPF", "PTTG1", "EXO1",
                 "ORC6L", "ANLN", "CCNE1", "CDC20", "MKI67",
                 "KIF2C", "ACTR3B", "MYC", "EGFR", "KRT5",
                 "PHGDH", "CDH3", "MIA", "KRT17", "FOXC1",
                 "SFRP1", "KRT14", "ESR1", "SLC39A6", "BAG1",
                 "MAPT", "PGR", "CXXC5", "MLPH", "BCL2",
                 "MDM2", "NAT1", "FOXA1", "BLVRA", "MMP11",
                 "GPR160", "FGFR4", "GRB7", "TMEM45B", "ERBB2")

samples.toc <- fread('data/samples-toc.tsv', header=TRUE)
samples.toc[WGS.project == "NA",    WGS.project := NA]
samples.toc[WTS.project == "NA",    WTS.project := NA]
samples.toc[vcf.file == "NA",       vcf.file := NA]
samples.toc[counts.file == "NA",    counts.file := NA]

read.counts <- fread('data/read-counts.txt', header=TRUE)
setkey(read.counts, 'project')
```

```{r, echo=FALSE}
counts.wide.all <- data.table()

load.counts <- function(fn, pid, sid) {
    col.nms <- c('ensembl.id', paste0(pid, '.', sid))
    fd <- fread(paste0('data/', fn), col.names=col.nms)
    setkey(fd, 'ensembl.id')
    if (nrow(counts.wide.all) == 0) {
        counts.wide.all[, ensembl.id := fd$ensembl.id]
        setkey(counts.wide.all, 'ensembl.id')
    }
    stopifnot(counts.wide.all$ensembl.id == fd$ensembl.id)
    counts.wide.all[, (col.nms[2]) := fd[, 2]]
}

unused <- samples.toc[!is.na(counts.file), mapply(load.counts, counts.file, patient.id, sample.id)]

counts.wide <- counts.wide.all[grepl("^ENS", ensembl.id)]

counts.long.all <- melt(counts.wide.all, id.vars='ensembl.id')[, .(ensembl.id, sample = variable, count = value)]
counts.long.all[, patient.id := gsub("[.].*", "", sample)]
counts.long <- counts.long.all[grepl("^ENS", ensembl.id)]
```

## Basic data summary

The following samples underwent Whole Transcriptome Sequencing (WTS):

```{r, echo=FALSE}
x <- samples.toc
setkey(x, 'WTS.project')
x <- merge(x, read.counts, by.x='WTS.project', by.y='project')
setkeyv(x, c('patient.id', 'sample.id'))
knitr::kable(x[!is.na(WTS.project), .(patient.id, sample.id, WTS.project, read.count = commaize(read.count), sample.site)], longtable=FALSE)
```

The reader were aligned to hg19 with STAR, using Homo\_sapiens.GRCh37.87.gtf.
Using `htseq-count` to count the number of reads per gene, we get the following raw counts:

```{r, echo=FALSE, results='asis'}
alignment.summary <- counts.long.all[,
    .(fragments = sum(count),
      good.count = sum(grepl("^ENS", ensembl.id) * count),
      good.frac = sum(grepl("^ENS", ensembl.id) * count)/sum(count),
      no.feature = sum((ensembl.id == "__no_feature") * count)/sum(count),
      ambiguous = sum((ensembl.id == "__ambiguous") * count)/sum(count),
      non.unique = sum((ensembl.id == "__alignment_not_unique") * count)/sum(count)),
    by=.(patient.id, sample)]

for (pid in unique(alignment.summary$patient.id)) {
    print(knitr::kable(alignment.summary[patient.id == pid,
        .(sample,
          fragments = commaize(fragments),
          good.count = commaize(good.count),
          good.frac = percentize(good.frac),
          no.feature = percentize(no.feature),
          ambiguous = percentize(ambiguous),
          non.unique = percentize(non.unique)
        )], caption=pid))
}
```

The raw counts have the following distribution:

```{r, echo=FALSE}
ggplot(counts.long[count > 1], aes(log2(count))) +
    geom_density(aes(group=sample, fill=sample), alpha=0.25) +
    theme_minimal() +
    facet_wrap(~patient.id)
```

## Highly Expressed Genes

There are a small number of genes with large counts. In order to compare them across samples,
we perform top-quartile normalisation after discarding small counts (i.e. < 10), to yield counts per million (cpm)

```{r, echo=FALSE, results='asis'}
quant.summary <- counts.long[count >= 10,
    .(q25 = quantile(count, 0.25),
      q50 = quantile(count, 0.5),
      q75 = quantile(count, 0.75),
      sum.q75 = sum(count * (count > quantile(count, 0.75)))),
    by=.(patient.id, sample)]
quant.summary[, scaling.factor := 1e6/sum.q75]
for (pid in unique(quant.summary$patient.id)) {
    print(knitr::kable(quant.summary[patient.id == pid, .(sample, q25, q50, q75, sum.q75, scaling.factor)], caption=pid))
}
```

```{r, echo=FALSE}
cpm.long <- counts.long[, .(patient.id, sample, ensembl.id, cpm = count * quant.summary[sample==sample, scaling.factor])]
lcpm.long <- cpm.long[, .(patient.id, sample, ensembl.id, lcpm = log2(cpm))]

ggplot(lcpm.long, aes(lcpm)) +
    geom_density(aes(group=sample, fill=sample), alpha=0.25) +
    theme_minimal() +
    facet_wrap(~patient.id)
```

Taking the most expressed 50 genes, we see the following relative expression across the different samples for each patient.

```{r, echo=FALSE, fig.align='center'}
# Pull out the list of most highly expressed genes
#
lcpm.of.interest <- lcpm.long[, .(lcpm = max(lcpm)), by=.(ensembl.id)]
setkey(lcpm.of.interest, 'lcpm')
high.genes = tail(lcpm.of.interest, 50)$ensembl.id

# Use them to extract a subset of the data
#
hgdt <- lcpm.long[ensembl.id %in% high.genes]
setkey(hgdt, 'ensembl.id')
hgdt <- merge(hgdt, gene.names, by.x='ensembl.id', by.y='ensembl.id')

# Order them by mean lcpm.
#
hg.order <- hgdt[, .(lcpm.mean = sum(lcpm)/length(lcpm)), by=.(name)]
setkey(hg.order, 'lcpm.mean')
hgdt[, name := factor(name, levels=hg.order$name)]

for (pid in unique(hgdt$patient.id)) {
    print(ggplot(hgdt[patient.id==pid], aes(name, lcpm, colour=sample)) +
                geom_point() +
                labs(x='gene', y='lcpm', colour='sample', title=pid) +
                theme_minimal() +
                theme(axis.text.x = element_text(size = 7, angle = 90, hjust = 1), legend.position = "bottom"))
}
```

Taking the 50 genes with the greatest absolute variance, we see the following relative expression across the different samples for each patient.


```{r, echo=FALSE, fig.align='center'}
# Pull out the list of most highly expressed genes
#
lcpm.of.interest <- lcpm.long[, .(lcpm = var(lcpm)), by=.(ensembl.id)]
setkey(lcpm.of.interest, 'lcpm')
high.genes = tail(lcpm.of.interest, 50)$ensembl.id

# Use them to extract a subset of the data
#
hgdt <- lcpm.long[ensembl.id %in% high.genes]
setkey(hgdt, 'ensembl.id')
hgdt <- merge(hgdt, gene.names, by.x='ensembl.id', by.y='ensembl.id')

# Order them by mean lcpm.
#
hg.order <- hgdt[, .(lcpm.var = var(lcpm)), by=.(name)]
setkey(hg.order, 'lcpm.var')
hgdt[, name := factor(name, levels=hg.order$name)]

for (pid in unique(hgdt$patient.id)) {
    print(ggplot(hgdt[patient.id==pid], aes(name, lcpm, colour=sample)) +
                geom_point() +
                labs(x='gene', y='lcpm', colour='sample', title=pid) +
                theme_minimal() +
                theme(axis.text.x = element_text(size = 7, angle = 90, hjust = 1), legend.position = "bottom"))
}
```
